plugins/backstage-plugin-coder/src/components/CoderProvider/CoderAuthProvider.tsx

```diff
@@ -152,50 +140,140 @@ export const CoderAuthProvider = ({ children }: CoderAuthProviderProps) => {
-          setAuthToken('');
-        },
-      }}
-    >
-      {children}
-    </AuthContext.Provider>
-  );
-};
+  return {
+    ...authState,
+    isAuthenticated: validAuthStatuses.includes(authState.status),
+    registerNewToken: newToken => {
+      if (newToken !== '') {
+        setAuthToken(newToken);
+      }
+    },
+    ejectToken: () => {
+      setAuthToken('');
+      window.localStorage.removeItem(TOKEN_STORAGE_KEY);
+      queryClient.removeQueries({ queryKey: [CODER_QUERY_KEY_PREFIX] });
+    },
+  };
+}
+
+type AuthFallbackState = Readonly<{
+  trackComponent: TrackComponent;
+  hasNoAuthInputs: boolean;
+}>;
+
+function useAuthFallbackState(): AuthFallbackState {
+  // Can't do state syncs or anything else that would normally minimize
+  // re-renders here because we have to wait for the entire application to
+  // complete its initial render before we can decide if we need a fallback UI
+  const [isMounted, setIsMounted] = useState(false);
+  useEffect(() => {
+    setIsMounted(true);
+  }, []);
+
+  // Not the biggest fan of needing to keep the two pieces of state in sync, but
+  // setting the render state to a simple boolean rather than the whole Set
+  // means that we re-render only when we go from 0 trackers to 1+, or from 1+
+  // trackers to 0. We don't care about the exact number of components being
+  // tracked - just whether we have any at all
+  const [hasTrackers, setHasTrackers] = useState(false);
+  const trackedComponentsRef = useRef<Set<string>>(null!);
+  if (trackedComponentsRef.current === null) {
+    trackedComponentsRef.current = new Set();
+  }
+
+  const trackComponent = useCallback((componentId: string) => {
+    // React will bail out of re-renders if you dispatch the same state value
+    // that it already has, and that's easier to guarantee since the UI state
+    // only has a primitive. Calling this function too often should cause no
+    // problems, and most calls should be a no-op
+    const syncTrackerToUi = () => {
+      setHasTrackers(trackedComponentsRef.current.size > 0);
+    };
+
+    trackedComponentsRef.current.add(componentId);
+    syncTrackerToUi();
+
+    return () => {
+      trackedComponentsRef.current.delete(componentId);
+      syncTrackerToUi();
+    };
+  }, []);
+
+  return {
+    trackComponent,
+    hasNoAuthInputs: isMounted && !hasTrackers,
+  };
+}
+
+/**
+ * Exposes auth state for other components, but has additional logic for spying
+ * on consumers of the hook.
+ *
+ * Caveats:
+ * 1. This hook should *NEVER* be exposed to the end user
+ * 2. All official Coder plugin components should favor this hook over
+ *    useEndUserCoderAuth when possible
+ *
+ * A fallback UI for letting the user input auth information will appear if
+ * there are no official Coder components that are able to give the user a way
+ * to do that through normal user flows.
+ */
+export function useInternalCoderAuth(): CoderAuth {
```

Comment:
Not sure if there are better names for `useInternalCoderAuth` and `useEndUserCoderAuth`, but I wanted the names to communicate that only one of them should be used for the internal plugin code, and the other one is strictly for the end user

Reply:
`__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED`

In seriousness, sounds like reasonable names to me.  Part of me wants to use `External` to parallel `Internal` but I feel like "external auth" already has a different meaning "end user" seems better.